lxd/storage/drivers/driver_pure_util.go

```diff
@@ -12,3 +49,186 @@ func newPureClient(driver *pure) *pureClient {
+	}
+
+	return body, nil
+}
+
+// request issues a HTTP request against the PureStorage gateway.
+func (p *pureClient) request(method string, path string, reqBody io.Reader, reqHeaders map[string]string, respBody any, respHeaders map[string]string) error {
+	var url string
+
+	// Construct the request URL.
+	if strings.HasPrefix(path, "/api") {
+		// If the provided path starts with "/api", simply append it to the gateway URL.
+		url = fmt.Sprintf("%s%s", p.driver.config["pure.gateway"], path)
+	} else {
+		// Otherwise, prefix the path with "/api/<api_version>" and then append it to the gateway URL.
+		// If API version is not known yet, retrieve and cache it first.
+		if p.driver.apiVersion == "" {
+			apiVersions, err := p.getAPIVersions()
+			if err != nil {
+				return fmt.Errorf("Failed to retrieve supported PureStorage API versions: %w", err)
+			}
+
+			// Use the latest available API version.
+			p.driver.apiVersion = apiVersions[len(apiVersions)-1]
+		}
+
+		url = fmt.Sprintf("%s/api/%s%s", p.driver.config["pure.gateway"], p.driver.apiVersion, path)
+	}
+
+	req, err := http.NewRequest(method, url, reqBody)
+	if err != nil {
+		return fmt.Errorf("Failed to create request: %w", err)
+	}
+
+	// Set custom request headers.
+	for k, v := range reqHeaders {
+		req.Header.Add(k, v)
+	}
+
+	req.Header.Add("Accept", "application/json")
+	if reqBody != nil {
+		req.Header.Add("Content-Type", "application/json")
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: shared.IsFalse(p.driver.config["pure.gateway.verify"]),
+			},
+		},
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Failed to send request: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	// Wrap unauthorized requests into an API status error.
+	if resp.StatusCode == http.StatusUnauthorized {
+		return api.StatusErrorf(http.StatusUnauthorized, "Unauthorized request")
+	}
+
+	// Overwrite the response data type if an error is detected.
+	if resp.StatusCode != http.StatusOK {
+		respBody = &pureError{}
+	}
+
+	// Extract the response body if requested.
+	if respBody != nil {
+		err = json.NewDecoder(resp.Body).Decode(respBody)
+		if err != nil {
+			return fmt.Errorf("Failed to read response body from %q: %w", path, err)
+		}
+	}
+
+	// Extract the response headers if requested.
+	if respHeaders != nil {
+		for k, v := range resp.Header {
+			respHeaders[k] = strings.Join(v, ",")
+		}
+	}
+
+	// Return the formatted error from the body
+	pureErr, ok := respBody.(*pureError)
+	if ok {
+		pureErr.StatusCode = resp.StatusCode
+		return pureErr
+	}
+
+	return nil
+}
+
+// requestAuthenticated issues an authenticated HTTP request against the PureStorage gateway. In case
+// the access token is expired, the function will try to obtain a new one.
+func (p *pureClient) requestAuthenticated(method string, path string, reqBody io.Reader, respBody any) error {
+	retries := 1
```

Comment:
what are the retries for?

Reply:
please add a comment, as it wasnt clear if this was because of some flaky endpoint